lib_test/scheduler_test.ml

```diff
@@ -26,60 +26,50 @@ let test_delay_cancel () =
   let elapsed = after -. before in
   assert_bool "elapsed_time1" (elapsed < 0.4)
 
+let timed_wait_callback ~msg ?(time_min = 0.) ?(eps=0.1) ?(time_max = 60.) f =
+  let rd, wr = Unix.pipe () in
+  let finally () =
+    Unix.close rd;
+    Unix.close wr
+  in
+  Fun.protect ~finally (fun () ->
+      let before = Unix.gettimeofday () in
+      let after = ref None in
+      let callback () =
+        after := Some (Unix.gettimeofday ());
+        let (_:int) = Unix.write_substring wr " " 0 1 in
+        ()
+      in
+      f callback;
+      let (_: bool) = Thread.wait_timed_read rd time_max in
```

Comment:
is it worth adding a sanity check here that write time is as we expect (i.e. false when after is None)

Reply:
Added:
```
diff --git c/lib_test/scheduler_test.ml w/lib_test/scheduler_test.ml
index 536c0fde3..1d9129ccf 100644
--- c/lib_test/scheduler_test.ml
+++ w/lib_test/scheduler_test.ml
@@ -41,11 +41,13 @@ let timed_wait_callback ~msg ?(time_min = 0.) ?(eps=0.1) ?(time_max = 60.) f =
         ()
       in
       f callback;
-      let (_: bool) = Thread.wait_timed_read rd time_max in
-      match !after with
-      | None ->
+      let ready = Thread.wait_timed_read rd time_max in
+      match ready, !after with
+      | true, None ->
+        Alcotest.fail "pipe ready to read, but after is not set"
+      | false, None ->
         Alcotest.fail (Printf.sprintf "%s: callback not invoked within %gs" msg time_max)
-      | Some t ->
+      | _, Some t ->
         let actual_minimum = min (t -. before) time_min in
         Alcotest.(check (float eps))
           (Printf.sprintf "%s: callback invoked earlier than expected" msg) time_min actual_minimum)
```